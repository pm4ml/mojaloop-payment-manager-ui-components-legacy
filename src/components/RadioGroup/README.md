## RadioGroup

Renders a set of radio buttons

### Proptypes

|   Name   |   Type   |                 Description                 | Optional | Values |   Default   |
|----------|----------|---------------------------------------------|----------|--------|-------------|
| id       | `string` | Describes the HTML id                       |          |        | `undefined` |
| label    | `string` | Describes the label                         |          |        | `undefined` |
| options  | `[{}]`   | Describes radio inputs labels and values    |          |        | `[]`        |
| value    | `string` | Describes the selected value                |          |        | `undefined` |
| disabled | `bool`   | Describes whether should render as disabled |          |        | `false`     |
| onChange | `func`   | Describes the _onchange_ function           |          |        | `undefined` |







