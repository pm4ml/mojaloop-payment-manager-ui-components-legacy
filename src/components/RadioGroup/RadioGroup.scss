@import '../../assets/styles/vars/shadow.scss';

/* MULTI SELECTION */
.input-radio {
  padding: 3px;
  display: flex;
  font-size: 14px;
  align-items: center;
}
.input-radio__option {
  &:not(:first-child) {
    margin-left: 10px;
  }
}
.input-radio__input {
  appearance: none;
  display: none;
  box-shadow: 0 1px 2px rgba(0, 0, 0, 0.05), inset 0px -15px 10px -12px rgba(0, 0, 0, 0.05);
  position: relative;

  + label {
    align-items: center;
    max-width: 100%;
    display: flex;
    position: relative;
    cursor: pointer;
    display: flex;
    flex-direction: row;

    &:hover {
      &:before {
        box-shadow: $shadow-size rgba(0, 0, 0, 0.13);
      }
    }

    &:before {
      border-radius: 11px;
      box-sizing: initial;
      flex: 0 0 20px;
      display: inline-block;
      margin-right: 5px;
      background: linear-gradient(#fafafa, #f4f4f4);

      border: 1px solid #cacbcc;
      content: '';
      width: 20px;
      height: 20px;
      box-shadow: 0px 0px 0px 0px rgba(0, 0, 0, 0.13);
      transition: all 0.2s ease-out;
    }
  }

  &.disabled {
    + label {
      color: #999;
      &:before {
        background: #eee;
        box-shadow: none;
      }
    }
  }

  &.disabled.checked {
    + label {
      &:before {
        border-color: #60d2fd;
        background: #60d2fd;
      }
    }
  }

  &.checked {
    + label {
      &:before {
        border-color: #00a2df;
        background: linear-gradient(0deg, #20c2ff, #00a2df);
      }
      &:after {
        content: '';
        box-sizing: content-box;
        box-sizing: initial;
        position: absolute;
        left: 7px;
        top: 7px;
        border-radius: 4px;
        background: #fff;
        box-shadow: 0px 1px 2px rgba(0, 0, 0, 0.2);
        width: 8px;
        height: 8px;
        animation: animateRadioContent 0.2s ease-out 1;
      }
    }
  }
  &.focused {
    + label {
      &:before {
        box-shadow: $shadow-size rgba(0, 175, 255, 0.5);
      }
    }
  }
}
.input-radio__label {
  margin-left: 4px;
  color: #333;
  cursor: pointer;
  &:hover {
    color: #666;
  }

  &.disabled {
    color: #999;
  }
}

@keyframes animateRadioContent {
  0% {
    opacity: 0;
    transform: scale(0.5, 0.5);
  }
  100% {
    opacity: 1;
    transform: scale(1, 1);
  }
}
