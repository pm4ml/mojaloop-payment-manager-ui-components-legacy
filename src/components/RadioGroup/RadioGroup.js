import './RadioGroup.scss';

import PropTypes from 'prop-types';
import React, { PureComponent } from 'react';

import * as utils from '../../utils/common';
import keyCodes from '../../utils/keyCodes';

class RadioGroup extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      value: this.props.value,
      focused: undefined,
    };
    this.onFocus = this.onFocus.bind(this);
    this.onBlur = this.onBlur.bind(this);
    this.onChange = this.onChange.bind(this);
    this.testKey = this.testKey.bind(this);

    this.selectSiblingRadio = this.selectSiblingRadio.bind(this);
  }
  componentDidUpdate(prevProps) {
    const { value } = this.props;
    if (value !== prevProps.value) {
      this.setState({ value });
    }
  }
  onChange(value, disabled) {
    if (this.props.disabled || disabled) return;

    this.setState({ value, focused: value });
    this.button.focus();

    if (this.props.onChange) {
      this.props.onChange(value);
    }
  }
  onFocus(e) {
    e.preventDefault();
    e.stopPropagation();
    if (this.state.focused === undefined) {
      this.setState(state => ({ focused: state.value }));
    }
  }
  onBlur() {
    this.setState({ focused: undefined });
  }
  testKey(e) {
    const { keyCode, shiftKey } = e;
    if (keyCode === keyCodes.KEY_TAB) {
      e.preventDefault();
      this.setState({ focused: undefined });
      utils.focusNextFocusableElement(this.button, !shiftKey);
      return;
    }
    if (keyCode === keyCodes.KEY_LEFT) {
      e.preventDefault();
      this.selectSiblingRadio(false);
      return;
    }
    if (keyCode === keyCodes.KEY_RIGHT) {
      e.preventDefault();
      this.selectSiblingRadio(true);
    }
  }
  selectSiblingRadio(next) {
    const { value } = this.state;
    const { options } = this.props;
    let nextIndex = options.map(o => o.value).indexOf(value) || 0;
    let found = false;

    while (!found) {
      nextIndex += next ? 1 : -1;
      if (nextIndex === options.length || nextIndex < 0) {
        found = true;
        break;
      }
      if (!options[nextIndex].disabled) {
        const currentValue = options[nextIndex].value;
        this.setState({ value: currentValue, focused: currentValue });
        this.onChange(currentValue, false);
        found = true;
        break;
      }
    }
  }
  render() {
    const { value, focused } = this.state;
    const { id, label, disabled, options } = this.props;

    const classNames = utils.composeClassNames(['mb-input', 'input-radio', disabled && 'disabled']);
    return (
      <div className={classNames} id={id}>
        {label && <span>{label}</span>}
        {options.map((option, idx) => (
          <Radio
            key={idx.toString()}
            id={`${id}-${idx}`}
            onClick={this.onChange}
            checked={value === option.value}
            focused={focused === option.value}
            label={option.label}
            value={option.value}
            disabled={option.disabled || disabled}
          />
        ))}
        <input
          ref={button => {
            this.button = button;
          }}
          type="button"
          className="mb-input__holder"
          onFocus={this.onFocus}
          onBlur={this.onBlur}
          onKeyDown={this.testKey}
          disabled={disabled}
        />
      </div>
    );
  }
}

RadioGroup.propTypes = {
  id: PropTypes.string,
  label: PropTypes.string,
  options: PropTypes.arrayOf(PropTypes.shape()),
  value: PropTypes.string,
  disabled: PropTypes.bool,
  onChange: PropTypes.func,
};
RadioGroup.defaultProps = {
  id: undefined,
  label: undefined,
  options: [],
  value: undefined,
  disabled: false,
  onChange: undefined,
};

const Radio = ({ id, onClick, checked, label, focused, value, disabled }) => {
  const optionClassName = utils.composeClassNames(['input-radio__option', checked && 'checked']);
  const inputClassName = utils.composeClassNames([
    'input-radio__input',
    checked && 'checked',
    disabled && 'disabled',
    focused && 'focused',
  ]);
  return (
    <div className={optionClassName} onClick={() => onClick(value, disabled)} role="presentation">
      <div id={id} className={inputClassName} />
      <label htmlFor={id}>
        <span>{label}</span>
      </label>
    </div>
  );
};

Radio.propTypes = {
  id: PropTypes.string,
  onClick: PropTypes.func,
  checked: PropTypes.bool,
  label: PropTypes.string,
  focused: PropTypes.bool,
  value: PropTypes.string,
  disabled: PropTypes.bool,
};
Radio.defaultProps = {
  id: undefined,
  onClick: undefined,
  checked: undefined,
  label: undefined,
  focused: undefined,
  value: undefined,
  disabled: false,
};
export default RadioGroup;
