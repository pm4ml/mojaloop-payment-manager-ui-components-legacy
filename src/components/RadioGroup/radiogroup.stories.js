/* eslint no-console: "off" */
import React, { useState } from 'react';

import RadioGroup from './RadioGroup';
import { createOptionalValidation, validate, vd } from '../../reduxValidation';

export default {
  title: 'RadioGroup',
  component: RadioGroup,
};

function optionMaker(item, index) {
  return {
    label: `option number ${index + 1}`,
    value: `value number ${index}`,
  };
}

const options = [
  {
    label: 'disabled',
    value: 'disabled',
    disabled: true,
  },
  ...[...Array(3)].map(optionMaker),
];

const { value } = options[2];

export const Default = () => (
  <div className="m5">
    <RadioGroup options={options} type="text" placeholder="Default" value={value} />
  </div>
);

export const WithDisabled = () => {
  return (
    <div className="m5">
      <RadioGroup disabled options={options} placeholder="All disabled" />
    </div>
  );
};

export const WithEvents = () => (
  <div className="m5">
    <RadioGroup
      options={options}
      placeholder="Events (console)"
      onChange={newValue => console.log('onChange', newValue)}
      onClick={() => console.log('onClick')}
      onKeyPress={() => console.log('onKeyPress')}
      onEnter={() => console.log('onEnter')}
      onBlur={() => console.log('onBlur')}
      onFocus={() => console.log('onFocus')}
    />
  </div>
);
