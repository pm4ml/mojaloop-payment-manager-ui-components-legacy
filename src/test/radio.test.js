import { mount, shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import React from 'react';

import RadioGroup from '../components/RadioGroup';

const options = [
  {
    label: 'label1',
    value: 'value1',
  },
  {
    label: 'label2',
    value: 'value2',
    disabled: true,
  },
  {
    label: 'label3',
    value: 'value3',
  },
  {
    label: 'label4',
    value: 'value4',
  },
  {
    label: 'label5',
    value: 'value5',
  },
];

it('renders the radiogroup', () => {
  const wrapper = shallow(<RadioGroup options={options} />);
  expect(wrapper.find('.input-radio')).toHaveLength(1);
});

it('renders the prop id', () => {
  const wrapper = shallow(<RadioGroup options={options} id="testRadioId" />);
  expect(wrapper.find('#testRadioId')).toHaveLength(1);
});

it('renders the prop label', () => {
  const wrapper = shallow(<RadioGroup options={options} label="test-radio" />);
  expect(wrapper.find('span')).toHaveLength(1);
  expect(wrapper.find('span').text()).toBe('test-radio');
});

it('renders the disabled state', () => {
  const wrapper = shallow(<RadioGroup options={options} disabled />);
  expect(wrapper.find('.input-radio').hasClass('disabled')).toBe(true);
});

it('renders the radio options', () => {
  const wrapper = mount(<RadioGroup options={options} />);
  expect(wrapper.find('.input-radio__option')).toHaveLength(5);
});

it('renders the radio options labels', () => {
  const wrapper = mount(<RadioGroup options={options} />);
  expect(
    wrapper
      .find('.input-radio__option label')
      .at(0)
      .text(),
  ).toBe('label1');
});

it('renders the selected radio option', () => {
  const wrapper = mount(<RadioGroup options={options} value="value4" />);
  expect(wrapper.find('.input-radio__input.checked')).toHaveLength(1);
  expect(wrapper.find('.input-radio__option.checked label').text()).toBe('label4');
});

it('triggers onChange when changed', () => {
  const mockEvent = jest.fn();
  const wrapper = mount(<RadioGroup options={options} onChange={mockEvent} />);
  expect(mockEvent).not.toHaveBeenCalled();
  wrapper
    .find('.input-radio__option')
    .at(4)
    .simulate('click');
  expect(mockEvent).toHaveBeenCalled();
});

it('Changes selected radio when using Arrow keys', () => {
  const mockEvent = jest.fn();
  const wrapper = mount(<RadioGroup options={options} value="value1" onChange={mockEvent} />);
  expect(mockEvent).not.toHaveBeenCalled();
  wrapper
    .find('.mb-input__holder')
    .at(0)
    .simulate('focus');
  expect(wrapper.find('.input-radio__input.focused')).toHaveLength(1);
  wrapper
    .find('.mb-input__holder')
    .at(0)
    .simulate('keyDown', { key: 'Arrow Right', keyCode: 39, which: 39 });
  expect(mockEvent).toHaveBeenCalled();
});

it('Does not select a disabled option', () => {
  const mockEvent = jest.fn();
  const wrapper = mount(<RadioGroup options={options} value="value1" onChange={mockEvent} />);
  wrapper
    .find('.input-radio__option')
    .at(1)
    .simulate('click');
  expect(mockEvent).not.toHaveBeenCalled();
});

// Snapshot

it('renders the checkbox correctly when multiple props are set', () => {
  const wrapper = shallow(<RadioGroup options={options} id="test-radio" />);
  expect(shallowToJson(wrapper)).toMatchSnapshot();
});
