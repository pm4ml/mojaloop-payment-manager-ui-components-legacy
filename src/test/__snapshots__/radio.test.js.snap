// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the checkbox correctly when multiple props are set 1`] = `
<div
  className="mb-input input-radio"
  id="test-radio"
>
  <Radio
    checked={false}
    disabled={false}
    focused={false}
    id="test-radio-0"
    key="0"
    label="label1"
    onClick={[Function]}
    value="value1"
  />
  <Radio
    checked={false}
    disabled={true}
    focused={false}
    id="test-radio-1"
    key="1"
    label="label2"
    onClick={[Function]}
    value="value2"
  />
  <Radio
    checked={false}
    disabled={false}
    focused={false}
    id="test-radio-2"
    key="2"
    label="label3"
    onClick={[Function]}
    value="value3"
  />
  <Radio
    checked={false}
    disabled={false}
    focused={false}
    id="test-radio-3"
    key="3"
    label="label4"
    onClick={[Function]}
    value="value4"
  />
  <Radio
    checked={false}
    disabled={false}
    focused={false}
    id="test-radio-4"
    key="4"
    label="label5"
    onClick={[Function]}
    value="value5"
  />
  <input
    className="mb-input__holder"
    disabled={false}
    onBlur={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    type="button"
  />
</div>
`;
